### PR TITLE
CoreUI 4.0 Migrate

### PR DESCRIPTION
**Pull request by Krank3n**
CoreUI 4.0  Migrate
React Update
Webpack Update
Apollo Update

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "@babel/plugin-transform-regenerator",
     "@babel/plugin-transform-runtime",
     [
-      "@babel/plugin-proposal-class-properties", { "loose": true }
+      "@babel/plugin-proposal-class-properties", { "loose": false }
     ]
```
Comment (jakxnz) [inline .babelrc:11]: Can this be simplified?

```
"@babel/plugin-transform-runtime",
"@babel/plugin-proposal-class-properties"
```

File: cypress/integration/Header/Header.spec.js [hunk 2]
```diff
@@ -44,6 +44,13 @@ describe('Header', () => {
     it('Should provide navigation', () => {
       cy.get('.header-nav__item').should('have.length.greaterThan', 0).should('be.visible');
     });
-  });
 
+    it('Should provide breadcrumbs', () => {
+      cy.get('.breadcrumb-item').should('have.length.greaterThan', 0).should('be.visible');
+    });
+
+    it('Breadcrumb name should be title case', () => {
```
Comment (jakxnz) [inline cypress/integration/Header/Header.spec.js:52]:
```suggestion
    it('Should display breadcrumbs with title case', () => {
```

File: cypress/integration/Resort/Resort.spec.js [hunk 3]
```diff
@@ -59,7 +60,7 @@ describe('Resort', () => {
   });
 
   it('Rating should have a value', () => {
-    cy.get('.resort-rating__value').children('.progress-bar').should('have.attr', 'aria-valuenow', '76');
+    cy.get('.rating__value').children('.progress').children('.progress-bar').should('have.attr', 'aria-valuenow', '76');
```
Comment (jakxnz) [inline cypress/integration/Resort/Resort.spec.js:63]: Seems odd to need to do `.children()` twice? Maybe `.within()` is what you're looking for?

File: package.json [hunk 4]
```diff
@@ -27,76 +27,77 @@
...
     "react-apollo": "^3.1.5",
-    "react-dom": "^16.14.0",
+    "react-dom": "^17.0.2",
     "react-intl": "^3.12.1",
-    "react-markdown": "^6.0.1",
-    "react-router": "^5.2.0",
-    "react-router-dom": "^5.2.0"
+    "react-markdown": "^6.0.3",
+    "react-router": "^5.2.1",
+    "react-router-dom": "^5.3.0"
```
Comment (jakxnz) [inline package.json:100]: ❤️

File: src/js/components/Footer/Footer.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import { useQuery } from '@apollo/react-hooks';
-import { CFooter, CImg } from '@coreui/react';
+import { CFooter, CImage } from '@coreui/react';
```
Comment (jakxnz) [inline src/js/components/Footer/Footer.js:2]: Finally, people have stopped using "img"!

File: src/js/components/Header/Header.js [hunk 6]
```diff
@@ -46,15 +49,15 @@ const HeaderNavItems = () => {
 
   return (
     filtered.map((item) => (
-      <CHeaderNavItem key={item.path} className="header-nav__item px-3">
-        <CHeaderNavLink className="header-nav__link" to={item.path}>
-          <CIcon content={item.icon} className="header-nav__icon mr-2" />
+      <CNavItem key={item.path} className="header-nav__item px-3">
+        <CNavLink className="header-nav__link" href={item.path}>
+          <CIcon icon={item.icon} className="header-nav__icon me-2" />
```
Comment (jakxnz) [inline src/js/components/Header/Header.js:54]: Interesting. What does `me` do? "Margin end" or something?

File: src/js/components/SubHeader/SubHeader.js [hunk 7]
```diff
@@ -1,14 +1,52 @@
...
+        active: index + 1 === array.length,
+      });
+      return currentPathname;
+    });
+    return breadcrumbs;
+  };
+
+  const breadcrumbs = getBreadcrumbs(currentLocation);
+  return (
+    <>
```
Comment (jakxnz) [inline src/js/components/SubHeader/SubHeader.js:33]: No more breadcrumb router?

File: src/scss/_custom.scss [hunk 8]
```diff
@@ -7,7 +8,22 @@ button {
   }
 }
 
+.btn:focus,
+.btn.active:focus,
+.btn-check:focus + .btn,
+.btn-check:checked + .btn:focus,
+.btn-check:active + .btn:focus,
+.show > .btn.dropdown-toggle:focus {
+  box-shadow: 0 0 0.6rem var(--cui-btn-shadow);
```
Comment (jakxnz) [inline src/scss/_custom.scss:17]: Our upgrade needed box shadows 😅?

**Review by jakxnz — CHANGES_REQUESTED** [review]
Looking pretty good so far. Will give it a practical test on the next reivew

File: src/stories/coreui/Buttons.stories.js [hunk 9]
```diff
@@ -93,8 +77,8 @@ export const Buttons = () => (
             <a href="https://icons.coreui.io/">CoreUI Icons</a>
             .
           </p>
-          {['normal', 'active', 'disabled'].map((state, index) => (
-            <CRow className="align-items-center mb-3" key={index}>
+          {['normal', 'active', 'disabled'].map((state, i) => (
+            <CRow className="align-items-center mb-3" key={i}>
```
[the remark this answers is not in the document]
Comment (Krank3n) [reply src/stories/coreui/Buttons.stories.js:81]: Yea thanks for the heads up on this.

File: src/stories/coreui/Buttons.stories.js [hunk 10]
```diff
@@ -108,11 +92,11 @@ export const Buttons = () => (
                   'info',
                   'light',
                   'dark',
-                ].map((color, index) => (
+                ].map((color, ind) => (
                   <CButton
                     className="me-2"
                     color={color}
-                    key={index}
+                    key={ind}
```
[the remark this answers is not in the document]
Comment (Krank3n) [reply src/stories/coreui/Buttons.stories.js:99]: yea... I took out 'ind' and replaced with index.

Still need two variables as they are nested and will conflict

File: .nvmrc [hunk 11]
```diff
@@ -0,0 +1 @@
+16.0.0
```
[the remark this answers is not in the document]
Comment (Krank3n) [reply .nvmrc:1]: Yep changed that thanks dude

**Review by jakxnz — APPROVED** [review]
Punch it!

**Comment by jakxnz** [comment]
[CoreUI 4.0 Migrate](https://trello.com/c/jLMpqPm5/70-coreui-40-migrate)

**Pull request merged**